### PR TITLE
feat: deterministic message hashing 

### PR DESCRIPTION
**Pull request by fryorcraken**
Ref #1208 
After checking with @jm-clius, we should not change the gossipsub message id just yet. This change will have to be coordinated across the implementations.

However, we can provide the MUID calculation logic for now. This will be first use by nwaku to remove store duplicates.

### CODE REVIEW

**Comment by fryorcraken** [comment]
Can only be merged once #1230 introduces the `meta` field.

File: package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
     "packages/interfaces",
     "packages/enr",
     "packages/core",
+    "packages/message-hash",
```
Comment (weboko) [inline package.json:11]: I think it should be added to `release-please` as well.

**Comment by fryorcraken** [comment]
Remove reviewers as PR is actually still draft

**Comment by fryorcraken** [comment]
Now ready for review

File: packages/message-hash/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { sha256 } from "@noble/hashes/sha256";
+import type { IProtoMessage } from "@waku/interfaces";
+import { concat, utf8ToBytes } from "@waku/utils";
+
+/**
+ * Deterministic Message Hashing as defined in
+ * [14/WAKU2-MESSAGE](https://rfc.vac.dev/spec/14/#deterministic-message-hashing)
+ */
+export function messageHash(
```
Comment (fryorcraken) [inline packages/message-hash/src/index.ts:9]: Interesting part: the deterministic message hash logic

File: packages/core/src/lib/relay/message_id.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import { sha256 } from "@noble/hashes/sha256";
+import { messageHash } from "@waku/message-hash";
+import { proto_message } from "@waku/proto";
+import { concat } from "@waku/utils";
+import debug from "debug";
+
+import { toProtoMessage } from "../to_proto_message.js";
+
+const log = debug("waku:relay:message-id");
+export function gossipsubMessageId(
```
Comment (fryorcraken) [inline packages/core/src/lib/relay/message_id.ts:11]: interesting part: the gossipsub message id calculation

File: packages/message-hash/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      contentTopicBytes,
+      message.meta,
+    ]);
+  } else {
+    bytesToHash = concat([
+      pubsubTopicBytes,
+      message.payload,
+      contentTopicBytes,
+    ]);
+  }
```
[the remark this answers is not in the document]
Comment (fryorcraken) [reply packages/message-hash/src/index.ts:31]: It holds the bytes that needs to be hashed. The bytes to hash. What's _apt_?
Could probably just call it `bytes`

File: packages/core/src/lib/relay/message_id.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  message: Message
+): Promise<Uint8Array> | Uint8Array {
+  const startTime = performance.now();
```
[the remark this answers is not in the document]
Comment (fryorcraken) [reply packages/core/src/lib/relay/message_id.ts:14]: It's a JS API to handle performance. Which matters here as the methods need to execute within 100ms.

File: packages/message-hash/package.json [hunk 6]
```diff
@@ -0,0 +1,108 @@
+{
+  "name": "@waku/message-hash",
```
Comment (weboko) [inline packages/message-hash/package.json:2]: Should be added to `release-please` map.

File: packages/message-hash/.prettierignore [hunk 7]
```diff
@@ -0,0 +1,4 @@
+build
+bundle
+dist
+node_modules
```
Comment (weboko) [inline packages/message-hash/.prettierignore:4]: Let's add CHANGELOG to ignore

**Comment by fryorcraken** [comment]
TODO: Check the API around the seen cache key and ensure MUID is used for it.

**Comment by fryorcraken** [comment]
Discussing with @jm-clius, we do not want to do the change on Relay at this stage. We will want to coordinate across implementations when we do it.

**Pull request merged**